File: src/retraite_notionnelle/donnees/population.py
"""La pyramide des âges : combien de gens, à quel âge, quelle année.

C'est la pièce que `docs/limites.md` déclarait manquante. Sans elle, passer d'un
droit individuel à un coût collectif obligeait à supposer toutes les générations
de même taille — ce que le baby-boom dément d'un tiers, et ce qui interdisait
purement et simplement de projeter.

Une seule source, et une seule méthodologie de part et d'autre de la frontière :
le scénario central des **projections de population 2026** de l'INSEE, dont
l'onglet ``population`` porte l'effectif au 1er janvier par âge détaillé et par
année, de 1962 à 2070. L'INSEE date lui-même la frontière — estimations jusqu'en
2023, projections ensuite — et le dépôt la reprend telle quelle : ce qui est
projeté entre au niveau ``estimee``, et cette fiabilité se propage jusqu'au
résultat affiché.

Deux séries en sont tirées, qui ne servent pas à la même chose :

* ``effectif(age, annee)`` compte des **retraités** — d'où le plancher à 50 ans,
  en deçà duquel aucune pension de droit direct n'est servie ;
* ``actifs(annee)`` compte les **20-64 ans**, et ne sert qu'à projeter le
  dénominateur d'une part de PIB.
"""

from __future__ import annotations

import csv
from pathlib import Path

from .chargement import Fiabilite, SerieAnnuelle, charger_serie_annuelle

#: Âge plancher de la série, et donc âge en deçà duquel ``effectif`` rend zéro.
#: Le cas type qui liquide le plus tôt part à 52 ans.
AGE_MINIMAL = 50


class Population:
    """Effectifs par âge et par année, observés puis projetés."""

    def __init__(self, racine: Path) -> None:
        macro = racine / "reference" / "macro"
        self._effectifs: dict[int, dict[int, float]] = {}
        self._fiabilites: dict[int, Fiabilite] = {}
        chemin = macro / "population_par_age.csv"
        with chemin.open(encoding="utf-8") as flux:
            lignes = (l for l in flux if not l.lstrip().startswith("#"))
            for ligne in csv.DictReader(lignes):
                annee, age = int(ligne["annee"]), int(ligne["age"])
                self._effectifs.setdefault(annee, {})[age] = float(ligne["effectif"])
                niveau = Fiabilite.depuis_texte(ligne["fiabilite"])
                courante = self._fiabilites.get(annee)
                self._fiabilites[annee] = (
                    niveau if courante is None else min(courante, niveau)
                )
        if not self._effectifs:
            raise ValueError(f"aucune ligne exploitable dans {chemin}")

        self.actifs: SerieAnnuelle = charger_serie_annuelle(
            macro / "population_active.csv", "effectif", nom="population_active"
        )
        self._annees = sorted(self._effectifs)
        self.premiere_annee = self._annees[0]
        self.derniere_annee = self._annees[-1]
        self.age_maximal = max(
            age for effectifs in self._effectifs.values() for age in effectifs
        )

    # -- accès ---------------------------------------------------------------

    def _annee_bornee(self, annee: int) -> int:
        """Année ramenée dans la plage publiée.

        La série commence en 1962, la dépense observée en 1959 : les trois
        premières années empruntent la pyramide de 1962. C'est une approximation
        assumée, et elle porte sur trois années dont la dépense pèse un demi
        pour cent de celle d'aujourd'hui.
        """
        return min(max(annee, self.premiere_annee), self.derniere_annee)

    def effectif(self, age: int, annee: int) -> float:
        """Effectif d'un âge une année donnée ; zéro hors de la plage d'âges."""
        return self._effectifs[self._annee_bornee(annee)].get(age, 0.0)

    def effectif_tranche(self, age_debut: int, age_fin: int, annee: int) -> float:
        """Effectif cumulé d'une tranche d'âges, bornes comprises."""
        effectifs = self._effectifs[self._annee_bornee(annee)]
        return sum(
            effectifs.get(age, 0.0) for age in range(age_debut, age_fin + 1)
        )

    def fiabilite(self, annee: int) -> Fiabilite:
        """Fiabilité de la pyramide d'une année.

        Hors de la plage publiée, la valeur est empruntée : elle ne peut donc
        pas valoir mieux qu'``estimee``.
        """
        if annee < self.premiere_annee or annee > self.derniere_annee:
            return Fiabilite.ESTIMEE
        return self._fiabilites[annee]

    def annees(self) -> list[int]:
        return list(self._annees)

File: moteur/js/population.js
/**
 * La pyramide des âges : combien de gens, à quel âge, quelle année.
 *
 * Portage de ``src/retraite_notionnelle/donnees/population.py``. C'est la pièce
 * qui manquait au dépôt pour passer d'un droit individuel à un coût collectif :
 * sans elle, il fallait supposer toutes les générations de même taille — ce que
 * le baby-boom dément d'un tiers, et ce qui interdisait toute projection.
 *
 * Une seule source de part et d'autre de la frontière : le scénario central des
 * projections de population 2026 de l'INSEE, observé jusqu'en 2023 et projeté
 * ensuite. Ce qui est projeté porte la fiabilité `estimee`, qui se propage
 * jusqu'au résultat affiché.
 */

import { Fiabilite, SerieAnnuelle } from "./serie.js";

/** Âge plancher de la série : aucune pension de droit direct avant. */
export const AGE_MINIMAL = 50;

/** Effectifs par âge et par année, observés puis projetés. */
export class Population {
  constructor(paquet) {
    const brut = paquet.population;
    this.annees = brut.annees;
    this.ages = brut.ages;
    this._effectifs = brut.effectifs;
    this._fiabilites = brut.fiabilites;
    this.actifs = SerieAnnuelle.depuisPaquet("population_active", brut.actifs);
    this.premiereAnnee = this.annees[0];
    this.derniereAnnee = this.annees[this.annees.length - 1];
    this.ageMinimal = this.ages[0];
    this.ageMaximal = this.ages[this.ages.length - 1];
    this._rangs = new Map(this.annees.map((annee, rang) => [annee, rang]));
  }

  /**
   * Année ramenée dans la plage publiée, et la borne haute REFUSE.
   *
   * EN DEÇÀ, ON EMPRUNTE : la série commence en 1962, la dépense observée en
   * 1959, et les trois premières années empruntent la pyramide de 1962 — trois
   * années dont la dépense pèse un demi pour cent de celle d'aujourd'hui.
   *
   * AU-DELÀ, ON REFUSE, et l'asymétrie est le sujet. Emprunter la pyramide de
   * 2070 pour 2085 ne décale pas une population de quinze ans : cela rend,
   * sous le nom des 85 ans de 2085, l'effectif des 85 ans de 2070, nés quinze
   * ans plus tôt. Une pyramide s'indexe par ÂGE, et reconduire un âge change
   * de cohorte — là où reconduire la valeur de bord d'une série annuelle ne
   * change qu'un niveau. Le chiffre emprunté est PLAUSIBLE, et c'est ce qui
   * rend le refus nécessaire : rien d'autre ne le signalerait. Une cohorte
   * déjà née se prolonge par sa propre survie.
   */
  _anneeBornee(annee) {
    if (annee > this.derniereAnnee) {
      throw new Error(
        `pyramide des âges demandée en ${annee}, au-delà de ${this.derniereAnnee} `
        + "que l'INSEE projette. Emprunter la dernière pyramide changerait de "
        + "cohorte sans le dire : une cohorte déjà née se prolonge par sa "
        + "propre survie.",
      );
    }
    return Math.max(annee, this.premiereAnnee);
  }

  /**
   * Effectif d'un âge une année donnée ; zéro hors de la plage d'âges. Refuse
   * au-delà de la dernière année projetée : voir `_anneeBornee`.
   */
  effectif(age, annee) {
    // Le refus vient AVANT le test d'âge : demander un âge hors plage en 2085
    // doit se heurter à la borne des années, pas rendre zéro comme si la
    // question avait un sens.
    const borne = this._anneeBornee(annee);
    if (age < this.ageMinimal || age > this.ageMaximal) {
      return 0.0;
    }
    const rang = this._rangs.get(borne);
    return this._effectifs[rang][age - this.ageMinimal];
  }

  /** Effectif cumulé d'une tranche d'âges, bornes comprises. */
  effectifTranche(ageDebut, ageFin, annee) {
    // Le refus est posé ICI et non délégué à `effectif` : une tranche vide
    // n'appellerait `effectif` sur aucun âge, et passerait sans rien dire.
    this._anneeBornee(annee);
    let somme = 0.0;
    for (let age = ageDebut; age <= ageFin; age += 1) {
      somme += this.effectif(age, annee);
    }
    return somme;
  }

  /**
   * Fiabilité de la pyramide d'une année. Hors de la plage publiée, la valeur
   * est empruntée : elle ne peut donc pas valoir mieux qu'`estimee`.
   */
  fiabilite(annee) {
    if (annee < this.premiereAnnee || annee > this.derniereAnnee) {
      return Fiabilite.ESTIMEE;
    }
    return this._fiabilites[this._rangs.get(annee)];
  }
}
